remote-config/src/fetch/shared.rs

```diff
@@ -0,0 +1,625 @@
+// Copyright 2021-Present Datadog, Inc. https://www.datadoghq.com/
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::fetch::{
+    ConfigApplyState, ConfigFetcher, ConfigFetcherState, ConfigInvariants, FileStorage, OpaqueState,
+};
+use crate::{RemoteConfigPath, Target};
+use std::collections::HashMap;
+use std::sync::atomic::{AtomicU32, AtomicU64, Ordering};
+use std::sync::{Arc, Mutex};
+use std::time::Duration;
+use tokio::select;
+use tokio::time::sleep;
+use tokio_util::sync::CancellationToken;
+use tracing::error;
+
+/// Fetcher which does a run-loop and carefully manages state around files, with the following
+/// guarantees:
+///  - A file at a given RemoteConfigPath will not be recreated as long as it exists I.e. it will
+///    always be drop()'ed before recreation.
+///  - It does not leak files which are no longer in use, i.e. it refcounts across all remote config
+///    clients sharing the same RefcountingStorage.
+///  - The state is always valid, i.e. there will be no intermittently expired files.
+pub struct SharedFetcher {
+    /// (env, service, version) tuple representing the basic remote config target
+    pub target: Arc<Target>, // could be theoretically also Mutex<>ed if needed
+    /// A unique runtime id. It must not be used by any other remote config client at the same
+    /// time. Is allowed to be changed at any time.
+    pub runtime_id: Arc<Mutex<String>>,
+    /// Each fetcher must have an unique id. Defaults to a random UUID.
+    pub client_id: String,
+    cancellation: CancellationToken,
+    /// Interval used if the remote server does not specify a refetch interval, in nanoseconds.
+    pub default_interval: AtomicU64,
```

Comment:
As mentioned in the review of the primary RFC loop - that info is for agents and not tracer clients so this can remain constant.  (I believe most tracers poll at 1s)

Reply:
Removed.